### ISSUE TITLE
Compilation on Windows fails

### ISSUE THREAD

**Issue by haesleinhuepf** [clEsperanto/CLIc_prototype#20]
Hi Stephane @StRigaud ,

I can't compile the  sources right away because of this error:
![image](https://user-images.githubusercontent.com/12660498/100454656-a3345480-30bd-11eb-821c-b9ee06e9371b.png)

Can this be fixed by setting some compiler flag in cmake?

I'll dig a bit deeper and let you know if I find a solution.

Cheers,
Robert

**Comment by StRigaud** [comment]
So this come from an OpenCL deprecation of the clCreateCommandQueue which does not exist anymore in newest version of OpenCL and has been replaced by clCreateCommandQueueWithProperties.   

Normally CMake should get the OpenCL version and depending on it, should trigger the proper function call.

I would advise you to clean and regenerate your CMake cache and project.
I will check this on my side also when I have 5 min

**Comment by haesleinhuepf** [comment]
No hurry. I can workaround it :-)

Cleaning doesn't help btw.

**Comment by StRigaud** [comment]
Just compiled the master branch on windows, no issue on my part.
Compilation done with CMake / VS2019 / CUDA SDK 11 (OpenCL 1.2) / Tiff 4.1.0

OpenCL_VERSION is normally define by CMake at generation time by `add_compile_definitions(OpenCL_VERSION=${OpenCL_VERSION_MAJOR})` and should correspond to 1 or 2 depending on your OpenCL.

**Comment by haesleinhuepf** [comment]
Hm. My system is very similar (CMake 3.19.1), VS2019, Tiff 4.1.0 - but I'm using Intel OpenCL SDK 2020.2.487.

If I wanted to add the OpenCL_VERSION manually? Where would I have to do this?

**Comment by StRigaud** [comment]
The least ugly way would be to replace `${OpenCL_VERSION_MAJOR}` by 1 or 2 depending on your OpenCL version.

The ugly way is to set a #define `OpenCL_VERSION_MAJOR = 1` on top of the cleCommandQueueManager.h.

The best would be to understand why CMake to not get the OpenCL version right ...

I will try again with Intel SDK

**Comment by tropf** [comment]
```
find_package(OpenCL REQUIRED)
```

does not set the required variables. Dump all cmake vars using

```
get_cmake_property(_variableNames VARIABLES)
foreach (_variableName ${_variableNames})
      message(STATUS "${_variableName}=${${_variableName}}")
endforeach()
```

No variable `OpenCL_VERSION` or `OpenCL_VERSION_MAJOR` is added by cmake on loading, and cmake then sets `#define OpenCL_VERSION` (empty constant), which then causes the error during preprocessing.

This is specified in the manual, as the version options are only set when `find_package` operates in *config mode*:

> When the [version] argument is given, Config mode will only find a version of the package that claims compatibility with the requested version (see format specification).
[from here](https://cmake.org/cmake/help/latest/command/find_package.html#version-selection)

This is outlined above on the same page:

> The command has two modes by which it searches for packages: “Module” mode and “Config” mode. The above signature selects Module mode. If no module is found the command falls back to Config mode, described below. This fall back is disabled if the MODULE option is given.

(tbh I just learned this too, so this is less of a RTFM but more of a WABFM: write a better fucking manual)

Whether this version information is set depends on how OpenCL was installed. Apparently on this windows installation, but also on my debian stable & testing, ubuntu 20.04 OpenCL was installed from the system.

On Linux the information on OpenCL is provided via `OpenCL.pc` (`find /usr -name 'OpenCL.pc'`), which contains the version information (`pkg-config --modversion OpenCL`) -- but apparently cmake does not load this information.

Suggested fix:
Retrieve version information via constants set by OpenCL itself, and **do not rely on cmake**.

**Comment by tropf** [comment]
checked back which constants OpenCL defines, implemented fix in #29 by using constant `CL_VERSION_2_0`
defined in [`CL/cl_version.h`](https://github.com/KhronosGroup/OpenCL-Headers/blob/master/CL/cl_version.h).

Note that `CL_VERSION_2_0` is set for version 2.0 **and above**, defined in this snippet from `CL/cl_version.h`:

```
#if CL_TARGET_OPENCL_VERSION >= 200 && !defined(CL_VERSION_2_0)
#define CL_VERSION_2_0  1
#endif
```
